[webkitpy] Executive.run_command should log command before running.

bls1999 opened the pull request:
#### 0e7432d919ef2a8c5e868ba22a23fce759355fbd
<pre>
[webkitpy] Executive.run_command should log command before running.
<a href="https://bugs.webkit.org/show_bug.cgi?id=302485">https://bugs.webkit.org/show_bug.cgi?id=302485</a>
<a href="https://rdar.apple.com/164649390">rdar://164649390</a>

Reviewed by Jonathan Bedard.

Executive.run_command only logs the command being run after it completes. When
debugging infrastructure issues, it helps to know what command was running and
when.

So, this changes the logging behavior of Executive.run_command such that it
logs what command is about to be run before it runs.

* Tools/Scripts/webkitpy/common/system/executive.py:
    (Executive.run_command): Change logging behavior.

Canonical link: <a href="https://commits.webkit.org/303002@main">https://commits.webkit.org/303002@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8365ed59c602b296a74fa2c32849ebd39b13cea8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130884 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3207 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41843 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138312 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/82545 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d6608d87-1022-47fb-89bb-0c17bc6de409) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132755 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3190 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3050 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/99722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/82545 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/21f39dd0-27d6-4d69-bbd8-65a26d114bef) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133830 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/117202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80416 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/14d27cf3-b8f6-4731-9f21-339bfcc5df30) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/130233 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/2220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/35331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/81569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/110809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/35834 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140792 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2951 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/2665 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/108236 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2997 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/113533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/108156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2257 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/31957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55988 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20376 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3019 "Built successfully") | | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/2840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/3040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2948 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/webkitpy/common/system/executive.py:
```diff
@@ -376,11 +376,13 @@ def run_command(self,
         stdin, string_to_communicate = self._compute_stdin(input)
         stderr = self.STDOUT if return_stderr else None
 
+        _log.debug(f'RUN_COMMAND: Running "{self.command_for_printing(args)}"...')
```
We shouldn't add `RUN_COMMAND:` here, these log lines already have quite a bit of metadata and our commands are not short.

gsnedders commented on Tools/Scripts/webkitpy/common/system/executive.py:
```diff
@@ -404,10 +408,11 @@ def run_command(self,
             exit_code = process.wait(timeout=Timeout.difference())
 
-            _log.debug('"%s" took %.2fs' % (self.command_for_printing(args), time.time() - start_time))
+            if exit_code or duration > 5:
```
Honestly I would rather we log this _regardless_ of duration.

For some of the frequent complaints about `git-webkit pr` performance, it's really death by a thousand small cuts rather than anything taking notably wrong.

We probably want to give different logging outputs when `exit_code` is and is not `None`.

bls1999 replied:
Both of these asks came from @JonWBedard -- I'm okay with logging twice in debug, I just want to make sure there's a consensus.

JonWBedard replied:
`git-webkit pr` doesn't actually use this codepath, for whatever that is worth, @gsnedders

gsnedders commented on Tools/Scripts/webkitpy/common/system/executive.py:
```diff
@@ -396,6 +398,8 @@ def run_command(self,
             else:
                 output = process.communicate(string_utils.encode(string_to_communicate, encoding='utf-8'), timeout=Timeout.difference())[0]
 
+            duration = round(time.monotonic() - start_time, ndigits=2)
```
I would suggest not rounding it here, rather instead only rounding at output-time.

There's also benefits to what the code used to do, including the time for the code below.

(Plus I think the `process.wait` is redundant below — once `process.communicate` has returned, we've either timed out (and aren't running this) or the process has terminated.)

bls1999 replied:
Sure, I can change it back to rounding (especially with your shorthand tip, that's really cool).

As for `process.wait`, I think that's populating `exit_code` more than doing anything functionally. I didn't touch it with my PR, so I'm not sure changing it is the best move here?

gsnedders replied:
To be clear, I'm not suggesting we change that in this PR, but we should tidy a lot of this up…

gsnedders commented on Tools/Scripts/webkitpy/common/system/executive.py:
```diff
@@ -376,11 +376,13 @@ def run_command(self,
         stdin, string_to_communicate = self._compute_stdin(input)
         stderr = self.STDOUT if return_stderr else None
 
+        _log.debug(f'Running "{self.command_for_printing(args)}"...')
```
Rather than calling `self.command_for_printing(args)` multiple times, we should just save the result in some variable; elsewhere I've gone with `repro_cmd`, but maybe in this case just `cmd_str` or something?

bls1999 replied:
Good point! I'll do that.

gsnedders commented on Tools/Scripts/webkitpy/common/system/executive.py:
```diff
@@ -404,10 +408,11 @@ def run_command(self,
 
         except subprocess.TimeoutExpired:
-            _log.debug('"%s" timed out after %.2fs' % (self.command_for_printing(args), time.time() - start_time))
+            _log.debug(f'"{self.command_for_printing(args)}" timed out after {round(time.monotonic() - start_time, ndigits=2)}s.')
```
```suggestion
            _log.debug(f'"{self.command_for_printing(args)}" timed out after {time.monotonic() - start_time:.2f}s.')
```

bls1999 replied:
That's a nifty shortcut! I'll forego the duration variable.

bls1999 committed:
[webkitpy] Executive.run_command should log command before running.
https://bugs.webkit.org/show_bug.cgi?id=302485
rdar://164649390

Reviewed by Jonathan Bedard.

Executive.run_command only logs the command being run after it completes. When
debugging infrastructure issues, it helps to know what command was running and
when.

So, this changes the logging behavior of Executive.run_command such that it
logs what command is about to be run before it runs.

* Tools/Scripts/webkitpy/common/system/executive.py:
    (Executive.run_command): Change logging behavior.

Canonical link: https://commits.webkit.org/303002@main
